notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "```python\n",
+    "decorated_solver = implicit_diff.custom_root(explicit_force_fn)(solver)\n",
+    "```\n",
+    "Here we have to use our ```explicit_force_fn```. It is also possible to use a different linear solver for the ```vjp``` computation using the ```solve``` argument of `implicit_diff.custom_root`. Jax comes with a handfull of sparse linear solvers in ```jax.scipy.sparse.linalg```."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "removed-completion",
+   "metadata": {},
+   "source": [
+    "# Adding neighbor lists and other auxiliary output to our solver"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "wound-nirvana",
+   "metadata": {},
+   "source": [
+    "Until now we have only worked with solvers that return one parameter `R_final`. Now we'll see how we have to change our `implicit_diff` function in order to be able to also have our solver return a neighbor list `nbrs` and the number of steps `num_steps` needed to reach the minimium.\n",
+    "\n",
+    "There is only one thing we have to change besides the normal stuff we have to change when using neighbor lists. We simply add the ```has_aux=True``` keyword to ```implicit_diff.custom_root```."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "massive-formation",
+   "metadata": {},
+   "source": [
+    "There's one more funny thing to note about the neighbor list version. Instead of getting an ```CustomVJPException``` when we forget to wrap our ```force_fn``` with ```lax.stop_gradient``` we get no explicit error message but we now get a memory leak instead which drastically slows down the computation. So pay attention that your are passing the correct version of your ```force_fn``` to your ```solver```!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "consecutive-adapter",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We add the parameter forced_rebuilding to make sure \n",
+    "# that it is possible to construct new neighbor lists during the optimization.\n",
+    "\n",
+    "def implicit_diff_nl(params,R_init,box_size,forced_rebuilding):\n",
+    "    neighbor_fn, energy_fn = energy.soft_sphere_neighbor_list(displacement, box_size, **params)\n",
+    "\n",
+    "    force_fn = jit(quantity.force(energy_fn))\n",
+    "    \n",
+    "    # wrap force_fn with a lax.stop_gradient to prevent a memory leak\n",
+    "    no_grad_force_fn = jit(lambda x,neighbor: lax.stop_gradient(force_fn(x,neighbor)))\n",
+    "    \n",
+    "    # make the dependence on the variables we want to differentiate explicit\n",
+    "    explicit_force_fn = jit(lambda R, neighbor, p: force_fn(R,neighbor,**p))\n",
+    "    \n",
+    "    def solver(params, x):\n",
+    "        # params are unused\n",
+    "        del params\n",
+    "        # need to use no_grad_force_fn!\n",
+    "        return run_minimization_while_nl(neighbor_fn,no_grad_force_fn, x, forced_rebuilding, shift, dt_start=0.001, dt_max=0.005)\n",
```

Comment:
For the purposes of the colab demo should we bake `forced_rebuilding`, `dt_start`, and `dt_max` into the `run_minimization_while_nl` function? It might make the code a bit easier to read.

Reply:
Done